staking-contract/src/lib.rs

```diff
@@ -0,0 +1,201 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_bindgen::{AccountId, Balance, BlockHeight, env, near_bindgen as near_bindgen_macro, Promise, PublicKey};
+use near_bindgen::collections::Map as NearMap;
+use std::str::FromStr;
+
+#[cfg(test)]
+mod test_utils;
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+const EPOCH_LENGTH: BlockHeight = 60;
+
+#[derive(BorshDeserialize, BorshSerialize, Debug)]
+pub struct User {
+    pub account_id: AccountId,
+    pub amount: Balance,
+    pub stake: Balance,
+    pub stake_height: BlockHeight,
+}
+
+impl User {
+    pub fn new(account_id: &AccountId, amount: Balance) -> Self {
+        Self {
+            account_id: account_id.clone(),
+            amount,
+            stake: 0,
+            stake_height: 0,
+        }
+    }
+
+    pub fn stake(&mut self, amount: Balance) {
+        self.amount -= amount;
```

Comment:
This can panic if `amount > self.amount`. Is it intentional, will be the error received by the user calling this function meaningful. (It happens few more time through the code).